docs/developer-docs/security/security-best-practices/data-integrity-and-authenticity.mdx

```diff
@@ -2,25 +2,628 @@
+
+Some examples of certified variables are asset certification in [Internet Identity](https://github.com/dfinity/internet-identity/blob/b29a6f68bbe5a49d048e12bc7a3263a9f43d080b/src/internet_identity/src/main.rs#L775-L808), [NNS dapp](https://github.com/dfinity/nns-dapp/blob/372c3562127d70c2fde059bc9c268e8ae858583e/rs/src/assets.rs#L121-L145), or the [canister signature implementation in Internet Identity](https://github.com/dfinity/internet-identity/blob/main/src/internet_identity/src/signature_map.rs).
+
+:::tip
+Certified variables is an advanced feature which require careful implementation of authenticated data structures and verification on the canister and client side respectively. **If the client doesn't require fast response times, call the query method as an update call (replicated query).** The response would be certified by the subnet and a single malicious or boundary node can't modify the response.
+:::
+
+:::tip
+ICP also provides replica signed queries, where query responses are signed by the answering replica node, however it doesn't have the same security guarantees an `update` call and only protects from malicious boundary nodes. Replica signed queries are enabled by default on both `agent-rs` and `agent-js`. [Read more about the feature](https://internetcomputer.org/blog/features/replica-signed-queries).
+:::
+
+### What is certified data?
+A canister can store a small amount of data (32 bytes) during an update call in the state tree under `certified_data` which will be certified by the subnet. During a query call, the canister can obtain a certificate about that data, which can be verified against the IC root public key. Combining these features, a canister should be able to certify a variable in a query response with this [design](https://medium.com/dfinity/how-internet-computer-responses-are-certified-as-authentic-2ff1bb1ea659). 
+
+On a high level, in your canister,
+1. Choose an [authenticated data strucutre](https://cs.brown.edu/research/pubs/pdfs/2003/Tamassia-2003-ADS.pdf) like Merkle trees for storing the value in canister memory.
+2. In the **update** call,
+  - Perform the computation and store the result in the Merkle store.
+  - The lookup path for the result must act as it's `key`. Ideally this `key` should be the parameters provided by the caller in the query method.
+  - Recompute the Merkle proof (`root_hash`)
+  - Store the `root_hash` as the canister's certified data.
+  - Return the `key` as response.
```

Comment:
Why do you need to return the key?

Reply:
The `key` would be used as the query parameter to fetch the data. If it's transparent to the user, the canister need not respond anything. But if you use an internal counter, I suppose you would need to return it as the response.